File: src/Tictac.cpp
//
// Created by LiZnB on 2023/12/23.
//

#include "Tictac.h"

Tictac::Tictac(int n) : chess(3) {}

long long Tictac::try_play(int x, int y) {
  if (board[x][y] != -1) return -1;
  long long t_hash = hash ^ zobrist[x][y][o] ^ ohash;
  return t_hash;
}

int Tictac::end() {
  for (int i = 0; i < 3; i++) {
    int ok = board[i][0];
    for (int j = 0; j < 3; j++) {
      if (board[i][j] != board[i][0] || board[i][j] == -1) ok = 2;
    }
    if (ok != 2) return ok;
  }
  for (int i = 0; i < 3; i++) {
    int ok = board[0][i];
    for (int j = 0; j < 3; j++) {
      if (board[j][i] != board[0][i] || board[j][i] == -1) ok = 2;
    }
    if (ok != 2) return ok;
  }
  if (board[0][0] == board[1][1] && board[1][1] == board[2][2] && board[0][0] != -1) return board[0][0];
  if (board[0][2] == board[1][1] && board[1][1] == board[2][0] && board[0][2] != -1) return board[0][0];
  for (int i = 0; i < 3; i++) {
    for (int j = 0; j < 3; j++) {
      if (board[i][j] == -1) return -1;
    }
  }
  return 2;
}

void Tictac::play(int x, int y) {
  hash ^= zobrist[x][y][o] ^ ohash;
  board[x][y] = o;
  o ^= 1;
}

void Tictac::pass() {
  o ^= 1;
  hash ^= ohash;
}

File: python/tictac.py
import numpy as np
import random
import mmh3


class Tictac:
    def __init__(self):
        self.n = 3
        self.o = 1
        self.o_hash = random.randint(0, 2 ** 32 - 1)
        self.board = np.full((3, 3), -1)

    def next(self):
        self.o ^= 1

    def play(self, x, y):
        if self.board[x, y] != -1:
            return False
        self.board[x, y] = self.o
        self.next()
        return True

    def state(self):
        for i in range(3):
            res = self.board[i, 0]
            for j in range(3):
                if self.board[i, j] != self.board[i, 0] or self.board[i, j] == -1:
                    res = 2        
            if res != 2:
                return res
            
        for i in range(3):
            res = self.board[0, i]
            for j in range(3):
                if self.board[j, i] != self.board[0, i] or self.board[j, i] == -1:
                    res = 2
            if res != 2:
                return res
        if np.all(self.board[[0, 1, 2], [0, 1, 2]] == self.board[1, 1]):
            return self.board[1, 1]
        if np.all(self.board[[0, 1, 2], [2, 1, 0]] == self.board[1, 1]):
            return self.board[1, 1]
        for i in range(3):
            for j in range(3):
                if self.board[i, j] == -1:
                    return -1
        return 2

    def hash(self):
        if self.o == 1:
            return mmh3.hash(self.board, signed=False) ^ self.o_hash
        else:
            return mmh3.hash(self.board, signed=False)

